source/linux/cpu_usage_sampler.c

```diff
@@ -0,0 +1,140 @@
+    double percent, total_combined;
+
+    uint64_t total_user_delta = 0, total_user_low_delta = 0, total_system_delta = 0, total_idle_delta = 0;
+    aws_get_cpu_sample_fn_linux_get_uint64_delta(total_user, sampler_linux->cpu_last_total_user, &total_user_delta);
+    aws_get_cpu_sample_fn_linux_get_uint64_delta(
+        total_user_low, sampler_linux->cpu_last_total_user_low, &total_user_low_delta);
+    aws_get_cpu_sample_fn_linux_get_uint64_delta(
+        total_system, sampler_linux->cpu_last_total_system, &total_system_delta);
+    aws_get_cpu_sample_fn_linux_get_uint64_delta(total_idle, sampler_linux->cpu_last_total_idle, &total_idle_delta);
+
+    total_combined = (double)(total_user_delta) + (double)(total_user_low_delta) + (double)(total_system_delta);
+    total = total_combined + (double)(total_idle_delta);
+
+    if (total == 0) {
+        *output = 0;
+        return_result = AWS_OP_ERR;
+        goto cleanup;
+    }
+
+    percent = (total_combined / total) * 100;
+
+    // If negative, there was an error (overflow?)
+    if (percent < 0) {
+        *output = 0;
+        return_result = AWS_OP_ERR;
+        goto cleanup;
+    }
+
+    *output = percent;
+    return_result = AWS_OP_SUCCESS;
+
+cleanup:
+    // Cache results
+    sampler_linux->cpu_last_total_user = total_user;
+    sampler_linux->cpu_last_total_user_low = total_user_low;
+    sampler_linux->cpu_last_total_system = total_system;
+    sampler_linux->cpu_last_total_idle = total_idle;
+
+    return return_result;
+}
+
+static void aws_cpu_sampler_destroy_linux(struct aws_cpu_sampler *sampler) {
+    if (sampler == NULL) {
+        return;
+    }
+    struct aws_cpu_sampler_linux *sampler_linux = (struct aws_cpu_sampler_linux *)sampler->impl;
+    aws_mem_release(sampler->allocator, sampler_linux);
+}
+
+static struct aws_cpu_sampler_vtable aws_cpu_sampler_vtable_linux = {
+    .aws_get_cpu_sample_fn = aws_get_cpu_sample_fn_linux,
+    .aws_cpu_sampler_destroy = aws_cpu_sampler_destroy_linux,
+};
+
+/*********************************************************************************************************************
+ * Public operations
+ ********************************************************************************************************************/
+
+struct aws_cpu_sampler *aws_cpu_sampler_new(struct aws_allocator *allocator) {
+    if (allocator == NULL) {
+        aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+        return NULL;
+    }
+
+    struct aws_cpu_sampler_linux *output_linux = aws_mem_calloc(allocator, 1, sizeof(struct aws_cpu_sampler_linux));
+    if (output_linux == NULL) {
+        return NULL;
+    }
+    output_linux->base.allocator = allocator;
+    output_linux->base.vtable = &aws_cpu_sampler_vtable_linux;
+    output_linux->base.impl = output_linux;
+
+    // CPU reporting is done via deltas, so we need to cache the initial CPU values
+    double tmp = 0;
+    aws_get_cpu_sample_fn_linux(&output_linux->base, &tmp);
```

Comment:
you could check for errors from the initial sample to see if our sampling code actually works on this platform.